torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+                gather_types(e, types, is_lhs);
+              }
+              return;
+            }
+            t = typeParser_.parseTypeFromExpr(expr);
+          } else {
+            lhs_val = emitExpr(expr);
+            t = lhs_val->type();
           }
-          return;
-        }
-        TypePtr type = typeParser_.parseTypeFromExpr(classinfo);
-        types.emplace_back(type);
-      }
-      bool staticallyTrue(const TypePtr& actual_type) {
-        // is this isinstance check statically true?
-        for (const TypePtr& typ : types) {
-          if (actual_type->isSubtypeOf(typ)) {
-            return true;
+
+          if (t->kind() == UnionType::Kind) {
+            auto inner = t->expect<UnionType>()->containedTypes().vec();
+            types.insert(types.end(), inner.begin(), inner.end());
+          } else if (t->kind() == OptionalType::Kind) {
+            auto inner = t->expect<OptionalType>()->getElementType();
+            types.push_back(inner);
+            types.push_back(NoneType::get());
+          } else if (t->kind() == NumberType::Kind) {
+            types.push_back(IntType::get());
+            types.push_back(FloatType::get());
+            types.push_back(ComplexType::get());
+          } else {
+            types.push_back(t);
           }
+        };
+
+    auto dedup = [&](std::vector<TypePtr>& types) {
+      std::sort(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) {
+            if (t1->kind() != t1->kind()) {
+              return t1->kind() < t2->kind();
+            }
+            return t1->str() < t2->str();
+          });
+
+      auto new_end = std::unique(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) { return *t1 == *t2; });
+
+      types.erase(new_end, types.end());
+    };
+
+    std::vector<TypePtr> lhs_types;
+    std::vector<TypePtr> rhs_types;
+
+    gather_types(obj, lhs_types, /*is_lhs=*/true);
+    gather_types(classinfo, rhs_types, /*is_lhs=*/false);
+
+    RefinementSet refinement;
+
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    if (obj.kind() == TK_VAR) {
+      std::string ident = Var(obj).name().name();
+
+      std::vector<Refinement> true_refinements;
+      std::vector<Refinement> false_refinements;
+
+      std::vector<TypePtr> isinstance_types;
+      std::vector<TypePtr> not_isinstance_types;
+
+      // We can discard any rhs types that we know statically would be
+      // impossible. For example, if we had:
+      //
+      //    def fn(x: Optional[str]):
+      //        if isinstance(x, (List[str], str, int)):
+      //            ...
+      //
+      // then `x` would be `str` in the true branch and `None` in the
+      // false branch, not `(List[str], str, int)` in the true branch
+      // and `None` in the false branch
+      for (TypePtr lhs_type : lhs_types) {
+        if (lhs_type == AnyType::get()) {
```

Comment:
I think we might be able to rewrite this to not special case Any if we iterated through the types on the rhs instead of the lhs, but i have to think about it more..

Reply:
This is evident that we're doing the logic wrong here. I think we want to check if lhs is a subtype of rhs or rhs is a subtype of lhs. However, we only insert refinements for the types in the rhs that subtype lhs.

Reply:
I think the only line that should be changed is `return lhs_type->isSubtypeOf(rhs_type)`, which should be `return rhs_type->isSubtypeOf(lhs_type)`. This definitely was a mistake, so thanks for catching it.

However, I don't understand why we'd want to check that `lhs` is a subtype of `rhs` or `rhs` is a subtype of `lhs`--wouldn't we introduce correctness issues if we allowed for "widening" refinements?

For example, consider:
```
def fn(x: Optional[int]):
    if isinstance(x, int):
        # "narrowing" refinement, aka the rhs is a subtype of the lhs
        # We know `x` is an int here and can use it as an int
```
in comparison to:
```
def fn(x: int):
    if isinstance(x, Optional[int]):
        # "widening" refinement, aka the lhs is a subtype of the rhs
        # This is bad! We already know `x` is an int; we shouldn't
        # insert casts so that it can now hold `None` instead
```